### PR TITLE
Introduce generation mode

### PR DESCRIPTION
**Pull request by chrfritsch**

### CODE REVIEW

File: src/Plugin/IdmsBuilder/RemoteBuilder.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    $replacedIdms = $this->replaceSnippetPlaceholders();
+    $files = $this->getAdditionalFiles();
+
+    /** @var \Drupal\thunder_print\Entity\PrintArticleInterface $printArticle */
+    $printArticle = $this->getPrintArticle();
+
+    $zip = new \ZipArchive();
+    $zipFilename = tempnam("tmp", "zip");
+
+    if ($zip->open($zipFilename, \ZipArchive::CREATE) !== TRUE) {
```
Comment (derhasi) [inline src/Plugin/IdmsBuilder/RemoteBuilder.php]: This should not fail silently.

File: src/Plugin/IdmsBuilder/RemoteBuilder.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      ->getBundleEntityType())
+      ->load($printArticle->bundle());
+
+    /** @var \Drupal\thunder_print\Entity\TagMappingInterface $tagMapping */
+    foreach ($bundle->getTags() as $tagMapping) {
+
+      /** @var \Drupal\Core\Field\FieldItemList $field */
+      $field = $printArticle->{$tagMapping->id()};
+
+      if (($fieldItem = $field->first()) && $tagMapping->getMappingType() instanceof MediaEntity) {
```
Comment (derhasi) [inline src/Plugin/IdmsBuilder/RemoteBuilder.php]: This should not be hardcoded on MediaEntity.

File: src/Plugin/IdmsBuilderBase.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+use Symfony\Component\HttpFoundation\File\MimeType\FileBinaryMimeTypeGuesser;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+
+/**
+ * Base class for Idms builder plugins.
+ */
+abstract class IdmsBuilderBase extends PluginBase implements IdmsBuilderInterface, ContainerFactoryPluginInterface {
+
+  protected $entityTypeManager;
```
Comment (derhasi) [inline src/Plugin/IdmsBuilderBase.php]: I miss docblocks here and below. Would provide better DX through autocompletion in IDE.

File: src/Plugin/TagMappingType/MediaEntity.php [hunk 4]
```diff
@@ -178,7 +181,7 @@ public function calculateDependencies() {
   /**
    * {@inheritdoc}
    */
-  public function replacePlaceholder(IDMS $idms, $fieldItem) {
+  public function replacePlaceholder(IdmsBuilderInterface $builder, IDMS $idms, $fieldItem) {
```
Comment (derhasi) [inline src/Plugin/TagMappingType/MediaEntity.php]: The TagMappingType should not need any knowledge of builder details. I would not inject the builder here, as it is much too specific.

I'd rather go with something like $buildMode being either '_singlefile_' or '_multifile_' or with providing an additional interface for TagMappingTypes, that can provide multiple files.

File: src/Plugin/IdmsBuilderBase.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  /**
+   * Get the print article.
+   *
+   * @return \Drupal\thunder_print\Entity\PrintArticleInterface
+   *   The print article.
+   */
+  protected function getPrintArticle() {
+    return $this->configuration['print_article'];
```
Comment (derhasi) [inline src/Plugin/IdmsBuilderBase.php]: The article should not be injected as configuration. The builder should be capable of building multiple articles without the need of getting reinitiated.

File: src/Plugin/IdmsBuilderBase.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    return $response;
+  }
+
+  /**
+   * Binary content that can be streamed.
+   *
+   * @return string
+   *   The content.
+   */
+  abstract protected function getContent();
```
Comment (chrfritsch) [inline src/Plugin/IdmsBuilderBase.php]: In Interface überführen

**Pull request merged**